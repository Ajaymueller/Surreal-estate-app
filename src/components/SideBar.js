import React, { useState } from 'react';
import '../styles/SideBar.css';
import qs from 'qs';
import { Link, useLocation, useHistory } from "react-router-dom";


const SideBar = () => {

    const history = useHistory();

    const { search } = useLocation();

    const [ query, setQuery ] = useState("");

    const buildQueryString = (operation, valueObj) => {
      
        const currentQueryParams = qs.parse(search, { ignoreQueryPrefix: true });
      
        const newQueryParams = {
          ...currentQueryParams,
          [operation]: JSON.stringify({
            ...JSON.parse(currentQueryParams[operation] || '{}'),
            ...valueObj,
          }),
        };
      
        return qs.stringify(newQueryParams, { addQueryPrefix: true, encode: false });
    };

    const handleSearch = (event) => {
        event.preventDefault();
        const newQueryString = buildQueryString('query', { title: { $regex: query }});
        history.push(newQueryString);
        setQuery("");
    };

    const handleChange = (e) => {
        setQuery(e.target.value);
    };

    return (
        <div className="side-bar">
            <form onSubmit={handleSearch}>
                <input type="text"
                value={query}
                onChange={handleChange} />
                <button type="submit">Search</button>
            </form>
        <ul className="navbar-links" data-testid="navbarlinks-id">
            <h3>Filter by city</h3>
            <Link to={`/?query={"city": "Manchester"}`}><li>Manchester</li></Link>
            <Link to={`/?query={"city": "Leeds"}`}><li>Leeds</li></Link>
            <Link to={`/?query={"city": "Sheffield"}`}><li>Sheffield</li></Link>
            <Link to={`/?query={"city": "Liverpool"}`}><li>Liverpool</li></Link>
            <h3>Sort by</h3>
            <Link to={buildQueryString('sort', { price: 1 })}><li>Ascending</li></Link>
            <Link to={buildQueryString('sort', { price: -1 })}><li>Descending</li></Link>
        </ul>
    </div>
    )
}

export default SideBar;