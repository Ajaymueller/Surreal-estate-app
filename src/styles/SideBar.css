.sidebar {
    grid-column: 2/ 4;
    grid-row: 2 / 3;
    font-size: 15px;
}

.navbar-links {
    text-decoration: none;
    list-style: none;
}