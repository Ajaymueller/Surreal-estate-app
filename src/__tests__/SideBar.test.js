import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import SideBar from '../components/SideBar';
import { MemoryRouter } from 'react-router-dom';
import "@testing-library/jest-dom/extend-expect";
import '@testing-library/jest-dom'

describe("sidebar", () => {
    it("renders correctly", () => {
        const { asFragment } = render(
        <MemoryRouter><SideBar /></MemoryRouter>)
        expect(asFragment()).toMatchSnapshot();
    });

    it("navbar links have option to filter by city and sort by", () => {
        const { getByText } = render(
        <MemoryRouter><SideBar /></MemoryRouter>)
        const filterByCity= getByText("Filter by city")
        const sortBy = getByText("Sort by")
        expect(filterByCity).toBeInTheDocument();
        expect(sortBy).toBeInTheDocument();
    })
});

describe("form", () => {
    it("should have an input", () => {
        const { getByTestId } = render(
            <MemoryRouter><SideBar/></MemoryRouter>)
        const input  = getByTestId("input-id")
        expect(input).toHaveClass("input");
    });
    it("input should update the state when changed", () => {
        const { getByTestId } = render(
            <MemoryRouter><SideBar/></MemoryRouter>)
        const input = getByTestId("input-id")
        fireEvent.change(input, { target: { value: "random title" }})
        expect(input.value).toBe('random title');
    });
    xit("calls handleSearch function on submit of form", () => {
        const handleSearch = jest.fn();
        const { getByTestId } = render(
            <MemoryRouter><SideBar/></MemoryRouter>)
        fireEvent.submit(getByTestId("form-id"));
        expect(handleSearch).toHaveBeenCalled();
    })
});;