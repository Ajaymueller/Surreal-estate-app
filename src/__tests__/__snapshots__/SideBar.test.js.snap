// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar renders correctly 1`] = `
<DocumentFragment>
  <div
    class="side-bar"
  >
    <form
      data-testid="form-id"
    >
      <label
        for="search-input"
      >
        <input
          class="input"
          data-testid="input-id"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Search
        </button>
      </label>
    </form>
    <ul
      class="navbar-links"
      data-testid="navbarlinks-id"
    >
      <h3>
        Filter by city
      </h3>
      <a
        href="/?query={\\"city\\": \\"Manchester\\"}"
      >
        <li>
          Manchester
        </li>
      </a>
      <a
        href="/?query={\\"city\\": \\"Leeds\\"}"
      >
        <li>
          Leeds
        </li>
      </a>
      <a
        href="/?query={\\"city\\": \\"Sheffield\\"}"
      >
        <li>
          Sheffield
        </li>
      </a>
      <a
        href="/?query={\\"city\\": \\"Liverpool\\"}"
      >
        <li>
          Liverpool
        </li>
      </a>
      <h3>
        Sort by
      </h3>
      <a
        href="/?sort={\\"price\\":1}"
      >
        <li>
          Ascending
        </li>
      </a>
      <a
        href="/?sort={\\"price\\":-1}"
      >
        <li>
          Descending
        </li>
      </a>
    </ul>
  </div>
</DocumentFragment>
`;
